parlai/utils/conversations.py

```diff
@@ -0,0 +1,188 @@
+            print(f'{k}: {v}')
+
+    @staticmethod
+    def _get_path(datapath):
+        fle, _ = os.path.splittext(datapath)
+        return os.path.join(fle, '.metadata')
+
+    @classmethod
+    def save_metadata(
+        cls,
+        datapath,
+        opt,
+        self_chat=False,
+        speaker_1='human',
+        speaker_2=None,
+        **kwargs,
+    ):
+        """
+        Dump conversation metadata to file.
+        """
+        metadata = {}
+        metadata['date'] = str(datetime.datetime.now())
+        metadata['opt'] = opt
+        metadata['self_chat'] = self_chat
+        metadata['speaker_1'] = speaker_1
+        if speaker_2 is not None:
+            metadata['speaker_2'] = speaker_2
+        else:
+            metadata['speaker_2'] = opt.get('model')
+
+        for k, v in kwargs.items():
+            metadata[k] = metadata[v]
+
+        metadata_path = cls._get_path(datapath)
+        print(f'[ Writing metadata to file {metadata_path} ]')
+        with open(metadata_path, 'w') as f:
+            f.write(json.dumps(metadata))
+
+
+class Conversations:
+    """
+    Utility class for reading and writing from ParlAI Conversations format.
+
+    Conversations should be saved in JSONL format, where each line is
+    a JSON of the following form:
+    {
+        'possible_conversation_level_info': True,
+        'dialogue':
+            [
+                {
+                    'id': 'speaker_1',
+                    'text': <first utterance>,
+                },
+                {
+                    'id': 'speaker_2',
+                    'text': <second utterance>,
+                },
+            ]
+        ...
+    }
+    """
+
+    def __init__(self, datapath):
+        self.conversations = self._load_conversations(datapath)
+        self.metadata = self._load_metadata(datapath)
+
+    @property
+    def num_conversations(self):
+        return len(self.conversations)
+
+    def _load_conversations(self, datapath):
+        if not os.path.isfile(datapath):
+            raise RuntimeError(
+                f'Conversations at path {datapath} not found. '
+                'Double check your path.'
+            )
+
+        conversations = []
+        with open(datapath, 'r') as f:
+            lines = f.read().splitlines()
+            for line in lines:
+                conversations.append(json.loads(line))
+
+        return conversations
+
+    def _load_metadata(self, datapath):
+        """
+        Load metadata.
+
+        Metadata should be saved at <identifier>.metadata
+        Metadata should be of the following format:
+        {
+            'date': <date collected>,
+            'opt': <opt used to collect the data,
+            'speaker_1': <identity of speaker 1>,
+            'speaker_2': <identify of speaker 2>,
+            ...
+            Other arguments.
+        }
+        """
+        try:
+            metadata = Metadata(datapath)
+            return metadata
+        except RuntimeError:
```

Comment:
should we also `except` other errors like `KeyError` if we're missing some fields in the MetaData?